### PR TITLE
Avoid allocations in prop rule key validation

### PR DESCRIPTION
**Pull request by djudd-stripe**
### Motivation
This saves ~5% on the `prop_definition` benchmark. (It looked a little better earlier and I think the impact will be bigger when we can remove the `valid_props` method entirely, which we can do after updating pay-server to override `valid_rule_key?` instead.)

### Test plan
Passing build

### CODE REVIEW

File: gems/sorbet-runtime/lib/types/props/decorator.rb [hunk 1]
```diff
@@ -61,23 +61,32 @@ def add_prop_definition(prop, rules)
     @props = @props.merge(prop => rules.freeze).freeze
   end
 
-  sig {returns(T::Array[Symbol])}
+  VALID_RULE_KEYS = %i{
```
Comment (DarkDimius) [inline gems/sorbet-runtime/lib/types/props/decorator.rb]: could you mark this constant private?

File: gems/sorbet-runtime/lib/types/props/optional.rb [hunk 2]
```diff
@@ -21,12 +21,14 @@ module T::Props::Optional::DecoratorMethods
     true,
   ].freeze
 
-  def valid_props
-    super + [
-      :default,
-      :factory,
-      :optional,
-    ]
+  VALID_RULE_KEYS = Set[
```
Comment (DarkDimius) [inline gems/sorbet-runtime/lib/types/props/optional.rb]: ditto, could we mark it private?

Comment (DarkDimius) [reply gems/sorbet-runtime/lib/types/props/optional.rb]: Also, if you're looking for most perf, Hash is faster than Set

File: gems/sorbet-runtime/lib/types/props/decorator.rb (hunk 1 above)
Comment (DarkDimius) [reply gems/sorbet-runtime/lib/types/props/decorator.rb]: Ditto about Hash being faster than Set

File: gems/sorbet-runtime/lib/types/props/optional.rb (hunk 2 above)
Comment (djudd-stripe) [reply gems/sorbet-runtime/lib/types/props/optional.rb]: Huh, that's silly. In a quick test looks like it's true specifically when you use `[]`, but `include?` is about the same either way, and maybe 50% slower - does that match your expectation? I guess `Hash#[]` has some special treatment?

Comment (djudd-stripe) [reply gems/sorbet-runtime/lib/types/props/optional.rb]: Made the constants private. I think I'll leave changing to Hash to when I'm able to delete `valid_props` entirely.

**Pull request merged**